### ISSUE TITLE
Rust runs 2fast2furious on the RPico?

### ISSUE THREAD

**Issue by JonasFocke01** [rust-lang/rust#118330]
Hello,

I encountered an error, where my code behaves differently in debug and release mode and i was not able to fix this with any black_box tricks. This can be, that im doing something wrong, but i banged my head at walls for hours not acheaving anyting.  
My case might be specific, but its rust non the less. Im programming a rpi pico and want to drive a [tm1637 chip](https://sming.readthedocs.io/en/latest/_inc/Sming/Libraries/TM1637/index.html) with it. Im using the tm1637 to drive a (seven segment [display](https://tinkersphere.com/2439/1355.jpg). I have boiled down my code to just the part that is not working in release mode. Here is the main part (without all the import stuff..):
```Rust
#[entry]
#[allow(clippy::too_many_lines)]
fn main() -> ! {
    #[allow(clippy::unwrap_used)]
    let mut pac = pac::Peripherals::take().unwrap();

    let mut watchdog = hal::Watchdog::new(pac.WATCHDOG);

    watchdog.start(ExtU32::secs(1));

    #[allow(clippy::unwrap_used)]
    let clocks = hal::clocks::init_clocks_and_plls(
        rp_pico::XOSC_CRYSTAL_FREQ,
        pac.XOSC,
        pac.CLOCKS,
        pac.PLL_SYS,
        pac.PLL_USB,
        &mut pac.RESETS,
        &mut watchdog,
    )
    .ok()
    .unwrap();

    let hal_sio = hal::Sio::new(pac.SIO);

    let pins = rp_pico::Pins::new(
        pac.IO_BANK0,
        pac.PADS_BANK0,
        hal_sio.gpio_bank0,
        &mut pac.RESETS,
    );
    let core = pac::CorePeripherals::take().unwrap();
    let mut delay = cortex_m::delay::Delay::new(core.SYST, clocks.system_clock.freq().to_Hz());

    let mut clock_pin = pins.gpio6.into_push_pull_output();
    let mut data_pin = pins.gpio7.into_push_pull_output();

    let tm = TM1637::new(&mut clock_pin, &mut data_pin, &mut delay);

    tm.init(); // <-- seems to work
   
    loop {
        tm.clear() // <-- works in dev, but not in release mode
        tm.display(&[5]) // <-- same behavior
    }
```
As you noticed, the TM1637 stuff is abstracted away, because it is a [crate](https://github.com/igelbox/tm1637-rs) that i used. I tinkered around with it, but i also cant find the issue there. I will not include its whole code (it could be anything), to not flood this issue, but here is a snapshot of it with the functions that _could_ be the problem here:
```Rust
    pub fn init(&mut self) -> Res<E> {
        self.start()?;
        self.send(ADDRESS_AUTO_INCREMENT_1_MODE)?;
        self.stop()?;

        Ok(())
    }

    pub fn clear(&mut self) -> Res<E> {
        self.print_raw_iter(0, core::iter::repeat(0).take(4))
    }

    pub fn print_raw(&mut self, address: u8, bytes: &[u8]) -> Res<E> {
        self.print_raw_iter(address, bytes.iter().map(|b| *b))
    }

    pub fn print_chars(&mut self, address: u8, chars: &[char]) -> Res<E> {
        self.print_raw_iter(address, chars.iter().map(|b| char_to_raw(*b)))?;
        Ok(())
    }

    pub fn print_hex(&mut self, address: u8, digits: &[u8]) -> Res<E> {
        self.print_raw_iter(
            address,
            digits.iter().map(|digit| DIGITS[(digit & 0xf) as usize]),
        )
    }

    pub fn print_raw_iter<Iter: Iterator<Item = u8>>(
        &mut self,
        address: u8,
        bytes: Iter,
    ) -> Res<E> {
        self.start()?;
        self.send(ADDRESS_COMMAND_BITS | (address & ADDRESS_COMMAND_MASK))?;
        for byte in bytes {
            self.send(byte)?;
        }
        self.stop()?;
        Ok(())
    }

    pub fn set_brightness(&mut self, level: u8) -> Res<E> {
        self.start()?;
        self.send(DISPLAY_CONTROL_BRIGHTNESS_BITS | (level & DISPLAY_CONTROL_BRIGHTNESS_MASK))?;
        self.stop()?;

        Ok(())
    }

    fn send(&mut self, byte: u8) -> Res<E> {
        let mut rest = byte;
        for _ in 0..8 {
            let bit = if rest & 1 != 0 { Bit::ONE } else { Bit::ZERO };
            self.send_bit_and_delay(bit)?;
            rest = rest >> 1;
        }

        // Wait for the ACK
        self.send_bit_and_delay(Bit::ONE)?;
        for _ in 0..255 {
            if self.dio.is_low()? {
                return Ok(());
            }
            self.delay();
        }

        Err(Error::Ack)
    }

    fn start(&mut self) -> Res<E> {
        self.send_bit_and_delay(Bit::ONE)?;
        self.dio.set_low()?;

        Ok(())
    }

    fn stop(&mut self) -> Res<E> {
        self.send_bit_and_delay(Bit::ZERO)?;
        self.dio.set_high()?;
        self.delay();

        Ok(())
    }

```
I encountered many errors, but i saw `Infallible` in release mode some times so i thought it would be best, to open an issue here. (also got pointed by reddit users).  
Can you spot anything that im doing wrong that could be the problem here? Or is this really a bug in the compiler or something like that?

**Comment by workingjubilee** [comment]
Please include the entire main.rs file. Do minimize it as much as possible, yes, but please, include a correct and compiling version. If you wish to hide something because it is very long, please familiarize yourself with the HTML [`<details>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) element. But there are multiple 3000-line files in the compiler, and many files host comments terminating in "...I think?"

In particular, this part you gave doesn't seem correct:
```rust
    let tm = TM1637::new(&mut clock_pin, &mut data_pin, &mut delay);

    tm.init(); // <-- seems to work
   
    loop {
        tm.clear() // <-- works in dev, but not in release mode
        tm.display(&[5]) // <-- same behavior
    }
```
This Rust code would emit syntax errors, as you need a `;` between `tm.clear()` and `tm.display`. In addition, those two functions take `&mut`, but you have declared `let tm`, instead of `let mut tm`.

There have been bugs regarding infinite loops in the past, partly because these are UB in some cases in C++, and LLVM is an optimizer which historically has been used to optimize C++ code, so there have been incidents where Rust code has been "optimized" according to C++ semantics. The code, at-a-glance, and looking past the issues I mentioned, seems like it ought to be sound, and given that, it should do what you expect it to in Rust. Thus it would be of great concern if this is indeed not working, as we might have to fix this upstream in LLVM. Of course, it might be a problem in the `unsafe` code in the crates you are using... this is the other reason it's very important that we *do* see all the imports.

I am curious if you have tried anything from `core::sync::atomic`, like [`compiler_fence`](https://doc.rust-lang.org/core/sync/atomic/fn.compiler_fence.html), to fix this?

**Comment by JonasFocke01** [comment]
Thanks for your effort!  
I have tinkered around and reduced my main file. This compiles and produces the error. I tested multiple times with it.  
Please let me know if something is still missing!  

```Rust
#![no_std]
#![no_main]

use panic_halt as _;
use rp_pico::entry;
use rp_pico::hal;
use rp_pico::hal::pac;
use rp_pico::hal::prelude::*;
use tm1637::TM1637;

#[entry]
fn main() -> ! {
    let mut pac = pac::Peripherals::take().unwrap();

    let mut watchdog = hal::Watchdog::new(pac.WATCHDOG);

    let clocks = rp2040_hal::clocks::init_clocks_and_plls(
        rp_pico::XOSC_CRYSTAL_FREQ,
        pac.XOSC,
        pac.CLOCKS,
        pac.PLL_SYS,
        pac.PLL_USB,
        &mut pac.RESETS,
        &mut watchdog,
    )
    .ok()
    .unwrap();

    let hal_sio = hal::Sio::new(pac.SIO);

    let pins = rp_pico::Pins::new(
        pac.IO_BANK0,
        pac.PADS_BANK0,
        hal_sio.gpio_bank0,
        &mut pac.RESETS,
    );

    let core = pac::CorePeripherals::take().unwrap();
    let mut delay = cortex_m::delay::Delay::new(core.SYST, clocks.system_clock.freq().to_Hz());

    let mut clock_pin = pins.gpio6.into_push_pull_output();
    let mut data_pin = pins.gpio7.into_push_pull_output();

    let mut tm = TM1637::new(&mut clock_pin, &mut data_pin, &mut delay);

    tm.init();
    tm.set_brightness(255);
    tm.clear();
    loop {
        tm.print_raw(0, &[0x77]);
    }
}
```

**Comment by workingjubilee** [comment]
Excellent, thank you! I wonder if this is somehow related to the usage of the `#![no_main]` feature? My intuition suggests we may need to make sure that the `#[entry]` isn't being given an LLVMIR annotation that would "allow" eliminating the loop, but there's many other possible gotchas in miscompilations.

**Comment by workingjubilee** [comment]
Objectively this _should_ be a low-priority issue if it does in fact require the nightly feature and is in fact limited to this platform. However, that labeling is somewhat "eager": it could be possible that the underlying bug actually _doesn't_ require that feature, and it could be a problem that is actually multiplatform. So it's worth further examination to confirm whether it is actually a higher-priority problem or not, because we've seen the much worse case of problem surface before.

**Comment by JonasFocke01** [comment]
Thank you for your time!  
If this is a low prio issue for you, thats totaly okay. I will try my best to check everything you mentioned.  
Two things stood out for me. First you say, that the code requires nightly. Where do you see that? I currently it says, that my active rustc version is `rustc version 1.74.0` installed. I cant find `nightly` in my project. Am i looking correct?  
Second, you speak about the loop getting optimized away. I dont know if that is it, because my bigger program does everything else in the loop as it should. Nothing breaks there.

**Comment by workingjubilee** [comment]
Oh, the answer is I confused the `#[entry]` annotation with `#[start]` and missed the `use rp_pico::entry` because I'm daft.

**Comment by workingjubilee** [comment]
> Second, you speak about the loop getting optimized away. I dont know if that is it, because my bigger program does everything else in the loop as it should. Nothing breaks there.

Hmm, that's interesting. So you're saying that in the "real case", there's huge blocks of code in a `loop {}`, but only the writes are getting optimized away? That points more strongly to a case of the library being UB in some subtle way, or the library's internal implementation being miscompiled.

**Comment by JonasFocke01** [comment]
Thats correct. And that something in that library gets compiled away was also my belief. I fought with the code once again, but i was not able to get it to work in release mode.  
The whole library thankfully is rather small. I strongly suspect, that it has something to do with the `send` function and the waiting for the acknowledge 'bit', but i was not able to get it to work by surrounding it with an endless loop, so my assumption could also be very wrong...  
Can you spot anything else in there (whole library)?  
```RUST
#![deny(warnings)]
#![no_std]

extern crate embedded_hal as hal;

use hal::blocking::delay::DelayUs;
use hal::digital::v2::{InputPin, OutputPin};

#[derive(Debug)]
pub enum Error<E> {
    Ack,
    IO(E),
}

impl<E> From<E> for Error<E> {
    fn from(err: E) -> Error<E> {
        Error::IO(err)
    }
}

type Res<E> = Result<(), Error<E>>;

pub struct TM1637<'a, CLK, DIO, D> {
    clk: &'a mut CLK,
    dio: &'a mut DIO,
    delay: &'a mut D,
}

enum Bit {
    ZERO,
    ONE,
}

impl<'a, CLK, DIO, D, E> TM1637<'a, CLK, DIO, D>
where
    CLK: OutputPin<Error = E>,
    DIO: InputPin<Error = E> + OutputPin<Error = E>,
    D: DelayUs<u16>,
{
    pub fn new(clk: &'a mut CLK, dio: &'a mut DIO, delay: &'a mut D) -> Self {
        Self { clk, dio, delay }
    }

    pub fn init(&mut self) -> Res<E> {
        self.start()?;
        self.send(ADDRESS_AUTO_INCREMENT_1_MODE)?;
        self.stop()?;

        Ok(())
    }

    pub fn clear(&mut self) -> Res<E> {
        self.print_raw_iter(0, core::iter::repeat(0).take(4))
    }

    pub fn print_raw(&mut self, address: u8, bytes: &[u8]) -> Res<E> {
        self.print_raw_iter(address, bytes.iter().map(|b| *b))
    }

    pub fn print_hex(&mut self, address: u8, digits: &[u8]) -> Res<E> {
        self.print_raw_iter(
            address,
            digits.iter().map(|digit| DIGITS[(digit & 0xf) as usize]),
        )
    }

    pub fn print_raw_iter<Iter: Iterator<Item = u8>>(
        &mut self,
        address: u8,
        bytes: Iter,
    ) -> Res<E> {
        self.start()?;
        self.send(ADDRESS_COMMAND_BITS | (address & ADDRESS_COMMAND_MASK))?;
        for byte in bytes {
            self.send(byte)?;
        }
        self.stop()?;
        Ok(())
    }

    pub fn set_brightness(&mut self, level: u8) -> Res<E> {
        self.start()?;
        self.send(DISPLAY_CONTROL_BRIGHTNESS_BITS | (level & DISPLAY_CONTROL_BRIGHTNESS_MASK))?;
        self.stop()?;

        Ok(())
    }

    fn send(&mut self, byte: u8) -> Res<E> {
        let mut rest = byte;
        for _ in 0..8 {
            let bit = if rest & 1 != 0 { Bit::ONE } else { Bit::ZERO };
            self.send_bit_and_delay(bit)?;
            rest = rest >> 1;
        }

        // Wait for the ACK
        self.send_bit_and_delay(Bit::ONE)?;
        for _ in 0..255 {
            if self.dio.is_low()? {
                return Ok(());
            }
            self.delay();
        }

        Err(Error::Ack)
    }

    fn start(&mut self) -> Res<E> {
        self.send_bit_and_delay(Bit::ONE)?;
        self.dio.set_low()?;

        Ok(())
    }

    fn stop(&mut self) -> Res<E> {
        self.send_bit_and_delay(Bit::ZERO)?;
        self.dio.set_high()?;
        self.delay();

        Ok(())
    }

    fn send_bit_and_delay(&mut self, value: Bit) -> Res<E> {
        self.clk.set_low()?;
        if let Bit::ONE = value {
            self.dio.set_high()?;
        } else {
            self.dio.set_low()?;
        }
        self.clk.set_high()?;
        self.delay();

        Ok(())
    }

    fn delay(&mut self) {
        self.delay.delay_us(DELAY_USECS);
    }
}

const MAX_FREQ_KHZ: u16 = 500;
const USECS_IN_MSEC: u16 = 1_000;
const DELAY_USECS: u16 = USECS_IN_MSEC / MAX_FREQ_KHZ;

const ADDRESS_AUTO_INCREMENT_1_MODE: u8 = 0x40;

const ADDRESS_COMMAND_BITS: u8 = 0xc0;
const ADDRESS_COMMAND_MASK: u8 = 0x0f;

const DISPLAY_CONTROL_BRIGHTNESS_BITS: u8 = 0x88;
const DISPLAY_CONTROL_BRIGHTNESS_MASK: u8 = 0x07;

const DIGITS: [u8; 16] = [
    0x3f, 0x06, 0x5b, 0x4f, //
    0x66, 0x6d, 0x7d, 0x07, //
    0x7f, 0x6f, 0x77, 0x7c, //
    0x39, 0x5e, 0x79, 0x71, //
];
```
and its config:
```TOML
[package]
name = "tm1637"
version = "0.1.0"
authors = [
    "Sergey Vakhurin <igelbox@gmail.com>",
    "Jesse Braham <jesse@beta7.io>",
]
description = "A platform agnostic driver to a LED-display powered by the TM1637 chip"
keywords = ["embedded", "embedded-hal-driver"]
categories = ["embedded", "hardware-support", "no-std"]
license = "MIT"
repository = "https://github.com/igelbox/tm1637-rs"
edition = "2018"
include = [
    "**/*.rs",
    "Cargo.toml",
]

[dependencies.embedded-hal]
features = ["unproven"]
version = "0.2.3"

[[example]]
name = "main"

[dev-dependencies]
embedded-hal = { version = "0.2.3" }
stm32f103xx-hal = { git = "https://github.com/japaric/stm32f103xx-hal" }
cortex-m = "0.6.2"
cortex-m-rt = "0.6.12"
panic-halt = "0.2.0"

[profile.release]
codegen-units = 1
debug = true
lto = true
opt-level = 'z'
```

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23118330.20Miscompilation.20of.20loop.20in.20rp_pico.20program)).

Please feel to raise the priority if the investigation uncovers a deeper issue in rustc/llvm

@rustbot label -I-prioritize +P-medium

**Comment by jannic** [comment]
I wonder if the program just gets too fast in release mode because of a missing delay in the `tm1637` library.
From that lib:
```
    fn send_bit_and_delay(&mut self, value: Bit) -> Res<E> {
        self.clk.set_low()?;
        if let Bit::ONE = value {
            self.dio.set_high()?;
        } else {
            self.dio.set_low()?;
        }
        self.clk.set_high()?;
        self.delay();

        Ok(())
    }
```
There is no delay between `self.clk.set_low()` and `self.clk.set_high()`. As setting a GPIO pin is infallible on rp2040, all the branches caused by the `?` operators get optimized away. So the resulting assembly becomes:
```
1000026c: 4a16          ldr     r2, [pc, #0x58]         @ 0x100002c8 <tm1637::TM1637<CLK,DIO,D>::send_bit_and_delay::hb6b8088b8d826102+0x60>
1000026e: 2340          movs    r3, #0x40
10000270: 6053          str     r3, [r2, #0x4]
10000272: 2900          cmp     r1, #0x0
10000274: d002          beq     0x1000027c <tm1637::TM1637<CLK,DIO,D>::send_bit_and_delay::hb6b8088b8d826102+0x14> @ imm = #0x4
10000276: 2180          movs    r1, #0x80
10000278: 6011          str     r1, [r2]
1000027a: e001          b       0x10000280 <tm1637::TM1637<CLK,DIO,D>::send_bit_and_delay::hb6b8088b8d826102+0x18> @ imm = #0x2
1000027c: 2180          movs    r1, #0x80
1000027e: 6051          str     r1, [r2, #0x4]
10000280: 6013          str     r3, [r2]
```
The `str rX, [r2, #0x4]` lines are the writes to `self.clk`. If I count cycles correctly, if `r1` is `0`, the time between clearing and setting the clock bit is only 5 cycles, or 40ns at 125MHz. That might be too fast for the hardware. And even if the TM1637 can handle such short clock pulses, it will not work reliably through a solderless breadboard or similar cabling.

You could try to reduce the CPU clock appropriately. That should make the clock pulse longer without changing anything else in your code. If that helps, it's a very strong hint that there's no miscompilation involved.

**Comment by jannic** [comment]
I found a datasheet for the TM1637 at https://raw.githubusercontent.com/avishorp/TM1637/master/docs/TM1637_V2.4_EN.pdf. According to that, the minimum clock pulse width should be 400ns. (Also, the maximum clock frequency should be 500kHz with 50% duty cycle, which would imply a minimum width of 1000ns.)
So an 48ns pulse is definitely too short.

To fix this issue in the tm1637 crate, I opened https://github.com/igelbox/tm1637-rs/pull/7/files

**TL;DR:** I don't think this is caused by a compiler bug.

**Comment by JonasFocke01** [comment]
@workingjubilee I definetly will! Im quire busy rn, we need to wait for the weekend

**Comment by workingjubilee** [comment]
Excellent, thank you very much with the help with that!
